### PR TITLE
Fire/Lava Fixes & Tweaks

### PR DESCRIPTION
**Pull request by ShizCalev**
:cl: ShizCalev
fix: Fixed held items on fire not being put out when the mob is extinguished.
fix: Corrected an inconsistency where items held by mobs on fire would not take damage when the rest of their equipment did.
fix: Fixed inconsistency where belts & bags would not be considered for burning when a mob is on fire.
fix: Flammable objects worn & held by a mob on fire will now properly ignite. Yes, you can now light candles by holding them while on fire!
fix: Fixed held items not igniting while standing over lava unless you reequipped them.
fix: Being immune to lava will no longer make your shoes/uniform/suit also immune to lava. Please make sure you properly lavaproof your equipment before swimming!
/:cl:

Added a helper to get all visible items a mob has equipped & holding. `mob/living/get_visible_items()`. Additional param `top_layer_only` will make it respect the `body_parts_covered` clothing flag.

### CODE REVIEW

**Comment by Poojawa** [comment]
Need to be able to make an ash drake backpack if normal ones burn now

**Comment by GunHog** [comment]
>fix: Being immune to lava will no longer make your clothing also immune to lava. Please make sure you properly lavaproof your equipment before swimming!

I believe that this warrants closer review, as it noticeable nerf to the lavaproof potion.

**Comment by Barhandar** [comment]
It's an _extreme_ nerf to all lavaproof clothes, not just the potion, that makes lavaproof loot (HECK, ash drake armor) _useless_. Also makes zero sense to have items in hands burn if you're lavaproof (whether you fixed the bug where taking items while on lava ignites them or not) because you're more or less on top of lava, not inside it. You've already added them catching on fire if you're on fire, that's enough.


Hence changelog is wrong, the last two "fixes" are actually nerfs. And the actual fixes are stealth buffs to lag because lag is #1 cause of running into lava and it will now fuck miners over _even more_ due to damaging their equipment even if they manage to extinguish themselves by resisting.

**Comment by ShizCalev** [comment]
Jaunters aren't effected since they're in a container object with the lavaproof flag. 

Fliers are already accounted for by lava/burn_stuff.

Borgs are unaffected by this change outside of having the cosmetic singeing effect applied to them. 

----------------------
Revised this to only burn all visible equipment if you're laying down on the lava.

> Also makes zero sense to have items in hands burn if you're lavaproof because you're more or less on top of lava, not inside it.

The current sources of lava immunity explicitly state when you use them that they allow you to walk -through- the lava, IE wading through it, not walk straight it. 

While immune to lava, only the clothing that come into contact with the lava (suit > uniform > shoes) and held items will be burned. Likewise, held items currently burn in live while lavaproof if you remove them from your bag, this is just fixing the inconsistency/bug with that. Backpacks & head items will no longer burn.

Now making your uniform & suit lavaproof will protect everything but what you have in-hand.

**Comment by ShizCalev** [comment]
@Barhandar `body_parts_covered` is an important part of what I said there. If you're wearing a lavaproofed suit, your shoes & uniform aren't burned. Gloves, head, and back items aren't burned in the current revision.

**Comment by Doppeltongue** [comment]
Is it just me, or is this a massive nerf to lava-proof golems?

**Comment by ishitbyabullet** [comment]
@Doppeltongue unless i read through the comments too fast, this wouldn't really affect golems unless they're holding an item, maybe.

Golems can't wear clothes and I don't think they get backpack/actual belts most of the time.

**Comment by ShizCalev** [comment]
@ishitbyabullet that is correct.

**Comment by Arianya** [comment]
@ShizCalev I can't be 100% but I think this somehow broke Play-Internet-Sound. Gonna try untestmerging it and seeing if it behaves correctly after that.

**Comment by zombiegums** [comment]
@ShizCalev  this change sets equipped borg modules on fire and will eventually destroy the module. 

image

My stun arm was the first item to disappear from my module inventory. I tried to put myself out while the items were equipped and it did nothing. Is this intended?

**Comment by ShizCalev** [comment]
@zombiegums Not intended, that leads off from the fact that the old way of dealing damage from burning was calling take_damage directly on the object instead of fire_act. I'll fix that in a few.

**Comment by Barhandar** [comment]
Fires set by lava ignore fireproofing added by potion. Miner in round 94911 had fireproofed _everything_ that can be fireproofed, plus speed potion on hardsuit, and their hardsuit still burnt down by lava.

This is _broken_.

File: code/game/turfs/simulated/lava.dm [hunk 1]
```diff
@@ -97,26 +104,15 @@
 	for(var/thing in thing_to_check)
 		if(isobj(thing))
 			var/obj/O = thing
-			if((O.resistance_flags & (LAVA_PROOF|INDESTRUCTIBLE)) || O.throwing)
```
Comment (ShizCalev) [inline code/game/turfs/simulated/lava.dm]: @Barhandar see here to from l100 - l109

**Comment by Barhandar** [comment]
>Unless you're saying you explicitly had FIREPROOFING, which isn't the same as LAVAPROOFING.

They had FIREPROOFING and died from FIRE that was ignited by lava.
>The behavior of lava removing fireproofing flags from objects if they aren't explicitly lavaproofed is how old lava worked. 

That's very dumb and illogical.
>The only difference now is that that same behavior also applies to worn objects in addition to just objects on the ground (again, that was an inconsistency in the handling of objects.)

And I think you're fixing it in the wrong way.

**Comment by HydroFloric** [comment]
I dont remember the round id but there is a issue with this when I was mining. As a angel flying *over* the lava taking items out of my bag would set them on fire which resulted in me losing a fully upgraded pka in the middle of a megafauna fight. 

Honestly the whole thing with held items being set on fire is fucking stupid since you would assume the fucking lava doesn't go up to your nipples. Even then you would think our characters are smart enough to hold them above the lava. 

Fix your fucking pr.

**Comment by HydroFloric** [comment]
Also another thing, items still light up if you hold them while on a fucking boat which makes just as much sense as flying over lava igniting objects. Once again

Fix your shit.

**Comment by ShizCalev** [comment]
@HydroFloric Neither of those are caused by this PR. It's due to storage handling's  `remove_from_storage()` chain calling `Entered()` on the turf you're on top of when you initially remove an item from storage.

@kevinz000 any ideas for that?

```
the lava (137,133,2) (/turf/open/lava): Entered(Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), the backpack (/obj/item/storage/backpack))
Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling): doMove(the lava (137,133,2) (/turf/open/lava))
Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling): doMove(the lava (137,133,2) (/turf/open/lava))
Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling): forceMove(the lava (137,133,2) (/turf/open/lava))
/datum/component/storage/concr... (/datum/component/storage/concrete): remove from storage(Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), the lava (137,133,2) (/turf/open/lava))
/datum/component/storage/concr... (/datum/component/storage/concrete): signal take obj(the backpack (/obj/item/storage/backpack), Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), the lava (137,133,2) (/turf/open/lava), 1)
/datum/callback (/datum/callback): InvokeAsync(the backpack (/obj/item/storage/backpack), Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), the lava (137,133,2) (/turf/open/lava), 1)
the backpack (/obj/item/storage/backpack): SendSignal("storage_take_obj", /list (/list))
Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling): attack hand(Iris Stocker (/mob/living/carbon/human))
Iris Stocker (/mob/living/carbon/human): UnarmedAttack(Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), 1)
Iris Stocker (/mob/living/carbon/human): ClickOn(Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), "icon-x=18;icon-y=17;left=1;scr...")
Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling): Click(null, "mapwindow.map", "icon-x=18;icon-y=17;left=1;scr...")
ShizCalev (/client): Click(Tapper: This One\'s For You (/obj/item/book/granter/action/drink_fling), null, "mapwindow.map", "icon-x=18;icon-y=17;left=1;scr...")
```

**Comment by HydroFloric** [comment]
@ShizCalev Ah sorry about that then, I swear I had never seen it happen before but its not like I ever went out of my way to test it. On a slightly related note is that a easy fix? Considering this is a lava/fire fix and tweak pr that seems like something that would be nice to be fixed.

**Comment by Barhandar** [comment]
The explorer's webbing still burns since it's a "belt" item even though it's technically and visually on the chest. Fixpls.

**Comment by Robustitron** [comment]
This breaks/nerfs fire protection in general.

Now your backpack melts off you, regardless of your fire protection, even atmos firesuits with full fire & heat immunity.

So now when atmos techs go off to fight a fire (or fight with fire), they have to first hide their backpack somewhere.

**Comment by Daxxed** [comment]
What the actual fuck ?
Not to even mention you claim golems can't get backpacks or belts.
Are we playing the same game ?

**Comment by ShizCalev** [comment]
@Robustitron See OP.

> Fixed inconsistency where belts & bags would not be considered for burning when a mob is on fire.

image
It seems strange that someone would assume that this

image
would keep this from catching on fire, especially when it normally burns when on the ground.

If there's something that's supposed to conceal the backpack however, let me know and I can fix that extremely easily.

**Comment by Barhandar** [comment]
>It seems strange that someone would assume that this would keep this from catching on fire, especially when it normally burns when on the ground.

It's called "unrealism for the sake of good gameplay", especially since YOU **CANNOT** FIREPROOF BACKPACKS AND OTHER CONTAINERS. The alternate solution would be making fire- and lavaproofing potions work on all items _and_ not just clothes and making engineering back items fireproof (stealth nerf to leather satchel option).

**Comment by HydroFloric** [comment]
Another thing related to the discussion, this will also be a stealth nerf to nukies who use the flamethrower kit(even though lets be honest nobody fucking uses that) and the elite hardsuit. 

since just like the atmos hardsuit its fireproof but not nukie backpacks. Which means some chucklefuck could go around lighting the syndies up and burning all their shit inside their bag.

**Comment by evsey9** [comment]
are you for fucking real? you're just making the atmos tech job obsolete and any job that requires you to go  into a fire because you'll lose all of your shit if you walk in the fire.

**Comment by Barhandar** [comment]
It's also a stealth buff to any and all flamethrowers.

**Comment by ShizCalev** [comment]
Actually looking at it, is the industrial backpack (engineers and atmosians) actually burning? It has the FIRE_PROOF flag already and shouldn't be.

**Comment by Robustitron** [comment]
I guess that flag is broken then, because every time I flamethrower a room I lost my atmos/engine backpack.

I wouldn't even mind this pr if atmos and pyro nukies had working fireproof backpacks

**Comment by ShizCalev** [comment]
#40798 fixed satchels and duffels. Will open another PR for syndie & ERT gear in a few. 

Also opened #40800 to held alleviate the issue of people not knowing stuff if something has special resistance attributes.

**Comment by XDTM** [comment]
Does a burning backpack delete all its contents upon being destroyed or does it spill them on the ground? It should be the latter, especially if some backpack contents are fireproof.

**Comment by Magikarpet** [comment]
I had a drake fight where my kinetic accelerator burnt 3 times and my backpack burnt 2 times from the fire breath and lava attacks. Having my jumpsuit melt off when i am wearing a firesuit/fireproof hardsuit seems silly, especially considering how crippling it is to lose your bag or your jumpsuit ingame. You either lose your tools to fight the fire, or you lose your ID, belt, or other pocket items.

I kinda feel like an ass for bringing it up but wiz's argument is fitting here.
https://forum.paradoxplaza.com/forum/index.php?threads/doesnt-development-end-very-unrealistic.872082/page-3#post-19679470
https://forum.paradoxplaza.com/forum/index.php?threads/why-common-sense-lacks-well-common-sense.868560/page-2#post-19595675
While this might be realistic, this change feels really awkward. A fire burning down your clothing slowly is a good mechanic but having all of your equipment burn down even when the fire is unable to damage you at all due to not being able to get a single item from xenobiology that takes 10-15 minutes to acquire is silly. It's even worse with lava as I don't think there is any way to lavaproof your items if slime potions aren't meant to protect equipment.


Though, while this does introduce a certain unique element or mechanic that the game lacks, and while I do like the concept of fires being able to burn down bags and other wearable items, it seems a bit silly due to the fact that most of the game doesn't work well around it. The only methods of earning protection are beyond scarce and limited and fire is a somewhat common threat on station, while lava is a permanent threat in mining. being an atmos technician and having all of your items except your hardsuit melt down, or mining and having all of your equipment disintegrate because you got hit three times by fire or stumbled into lava once. Realism doesn't feel like a fitting reason when an overwhelming majority of the game is not realistic.

tl;dr fire is too common and the protection items are too rare, it makes a decent chunk of jobs miserable to play because all of your weaponry or gear instantly melts off and burns even while you are mostly resistant to fire and not taking damage from it

but @XDTM it drops all of the contents

**Comment by XDTM** [comment]
The part where lava permanently removes any form of fire protection seems far from realistic in my opinion. 
Perhaps fire resistance could be expressed as an integer instead of a flag, so you could have varying grades of fire immunity; or even better some sort of direct damage reduction so fireproof items have an advantage over flammable ones when exposed to lava.

**Comment by ShizCalev** [comment]
@XTDM it stills the contents on the ground.

**Comment by ShizCalev** [comment]
https://github.com/tgstation/tgstation/blob/f4acdef7b9e871a275c65993c01dd7b6457e7cb9/code/modules/mob/living/carbon/human/life.dm#L120-L129
https://github.com/tgstation/tgstation/blob/00a8489c3d892fa9c01218c4e3b3535a5df90eeb/code/modules/mob/living/carbon/human/species.dm#L1631-L1639
https://github.com/tgstation/tgstation/blob/c8ca0c5d74a18719c575180a9493d1371fa0b0d3/code/game/objects/items.dm#L43

**Comment by XDTM** [comment]
That is the amount of heat protection they offer to the wearer, not their own fire protection. I meant replacing FIRE_PROOF and LAVA_PROOF with a fire damage reduction of some sort.

**Comment by ShizCalev** [comment]
Fire damage reduction should already be handled by the item's "fire" armor value (I think.)

**Comment by ShizCalev** [comment]
Also fixed xeno inhand items not catching on fire, as well as inactive cyborg modules catching on fire.

**Comment by XDTM** [comment]
That handles percentage reduction, but if you set it to 100 it becomes immune no matter the heat, which defies the point of a damage threshold that could be set to be immune to fire but only reduces lava.
Plus, as you coded it right now, fire resistance is cut down to 50 after coming in contact with lava, so you cannot lavaproof an item with it.

**Comment by Robustitron** [comment]
While we're here talking about fire fixes & consistency, I thought I'd mention that bug [#39766](https://github.com/tgstation/tgstation/issues/39766) is still a thing.

**Comment by adamkad1** [comment]
congratz, now items get set on fire and dont extinguish when you do. also firesuits can burn. what.

**Comment by ShizCalev** [comment]
@adamkad1 firesuit behavior hasn't been modified.

**Comment by HydroFloric** [comment]
just wanna say this does indeed cause fireproof 40k hardsuits to burn off.

**Comment by ShizCalev** [comment]
Testmerge on the server is using an old/conflicted version. Checking some stuff here to make sure nothing else broke with it.

**Comment by ShizCalev** [comment]
Gonna break atomize this further.
